protocol/network/4-architecture.md

```diff
@@ -0,0 +1,126 @@
+4 Architecture
+==============
+
+libp2p was designed around the Unix Philosophy, creating smaller components, easier to understand and to test. These components should also be able to be swapped in order to accomodate different technologies or scenarios and also make it that it is upgradable over time.
+
+Although different Peers can support different protocols depending on their capabilities, any Peer can act as a dialer and/or a listener for connections from other Peers, connections that once established can be reused from both ends, removing the distinction between clients and servers.
+
+libp2p interface acts as a thin veneer to three subsystems that are required in order for peers to be able to communicate. These subsystems are allowed to be built on top of other subsystems as long as they respect the standardized interface. The main 3 subsystems are:
+
+- Peer Routing - Mechanism to find a Peer in a network. This Routing can be done recursively, iteratively or even in a broadcast/multicast mode.
+- Swarm - Handles everything that touches the 'opening a stream' part of libp2p, from protocol muxing, stream muxing, NAT Traversal, Connection Relaying, while being multitransport
+- Distributed Record Store - A system to store provider records and relay records
+
+Each of these subsystem exposes a well known interface (see chapter 6 for Interfaces) and may use eachother in order to fulfil their goal. A global overview of the system is:
+
+```
+┌─────────────────────────────────────────────────────────────────────────────────┐
+│                                  libp2p                                         │
+└─────────────────────────────────────────────────────────────────────────────────┘
+┌─────────────────────────┐┌──────────────────────────┐┌──────────────────────────┐
+│       Peer Routing      ││      Swarm               ││ Distributed Record Store │
+└─────────────────────────┘└──────────────────────────┘└──────────────────────────┘
+```
+
+## 4.1 Peer Routing
+
+A Peer Routing subsystem, exposes an interface to identify which peers should a message be routed in the DHT. It receives a key and must return one or more PeerInfo objects.
+
+We present two examples of possible Peer Routing subsystems, the first based on a the Kademlia DHT and the second based on mDNS. Nevertheless, other Peer Routing mechanisms might be implemented, as long as they fulfil the same expectation and interface.
+
+### 4.1.1 kad-routing
+
+kad-routing implements the Kademlia Routing table, where each peer holds a set of k-buckets, each of them containing several PeerInfo from other peers in the network. In order to find the whereabouts of these peers, it implements 3 discovery mechanisms:
+
+- mDNS-discovery
+- random-walk
+- bootstrap-list
```

Comment:
yeah makes sense to list that it _can use_ them, but doesn't make sense for it to _define_ them, as a reader would think these are just _parts_ of kad-routing instead of general things, or that _kad-routing_ always (or most of the time) uses them. in reality, many ipfs nodes out there don't use mDNS. lots of people turn it off because VPSes dont like it